rfcs/20190814-kernel-and-op-registration.md

```diff
@@ -0,0 +1,658 @@
+  TF_OpDefinitionBuilderAddInput(b, "input: T");
+  TF_OpDefinitionBuilderAddOutput(b, "output: type");
+  TF_OpDefinitionBuilderAddAttr(b, "T: {bfloat16, ...}");
+  TF_OpDefinitionBuilderAddAttr(b, "type: {bfloat16, ...}");
+  TF_OpDefinitionBuilderSetShapeInferenceFunction(b, &InferBitcastShape);
+
+  TF_Status* status = TF_NewStatus();
+  TF_RegisterOpDefinition(b, status);
+  if (TF_GetCode(status) != TF_OK) { /* handle errors */ }
+}
+
+```
+
+### Shape Inference
+
+A significant feature of certain ops is their ability to infer their output
+shapes. TensorFlow will invoke the registered shape inference function (if one
+is provided) when it needs to know the op's output shape. The registration
+function declaration is shown below:
+
+
+```c++
+void TF_OpDefinitionBuilderSetShapeInferenceFunction(
+  TF_OpDefinitionBuilder* builder,
+  void (*shape_inference_func)(TF_ShapeInferenceContext* ctx, TF_Status* status));
+```
+
+A series of functions prefixed with `TF_ShapeInferenceContext` is provided for
+the following purposes:
+
+* Examining operator input shapes (`TF_ShapeInferenceContextGetInput`)
+
+* Creating and deleting shape and dimension handles (`TF_{New,Delete}ShapeHandle`, `TF_{New,Delete}DimensionHandle`)
+
+* Manipulating shape and dimension handles (`TF_ShapeInferenceContextWithRank`, `TF_ShapeInferenceContextDim`)
+
+In general, C analogues to the C++ methods in `tensorflow::shape_inference`
+(see `tensorflow/core/framework/shape_inference.h`) will be provided.
+
+## Kernels
+
+This section introduces changes to the C API that are required to support
+kernels. An alpha version of this API is already checked in at
+`tensorflow/c/kernels.h`.
+
+### Registration
+
+Kernel registration with the C++ API is accomplished with the
+`REGISTER_KERNEL_BUILDER` macro. This macro expands to code that relies on
+static initialization to register the provided kernel with the global kernel
+registry. See below for an example of registering a kernel with the C++ API:
+
+```c++
+
+#include "tensorflow/core/framework/op_kernel.h"
+
+class BitcastOp : public OpKernel {
+  explicit BitcastOp(OpKernelConstruction* context) : OpKernel(context) { … }
+  void Compute(OpKernelContext* context) override { … }
+};
+
+REGISTER_KERNEL_BUILDER(Name("Bitcast").Device(DEVICE_CPU), BitcastOp)
+```
+
+The equivalent C API provides a series of functions that operate on
+`TF_KernelBuilder`, an opaque struct obtained with the `TF_NewKernelBuilder` call.
+The kernel builder is registered with TensorFlow using the
+`TF_RegisterKernelBuilder` function. See below for an example of registering
+the bitcast kernel using the C API:
+
+```c++
+#include "tensorflow/c/kernels.h"
+
+typedef struct bitcast_kernel { … } bitcast_kernel;
+
+// Bitcast_Create, Bitcast_Compute and Bitcast_Delete actually implement the
+// kernel. See the section below for discussion on kernel implementation.
+static void* Bitcast_Create(TF_OpKernelConstruction* context) {
```

Comment:
In TF today op kernel construction can fail (and if often does, as it validates attributes etc). This should have a status somewhere.

Reply:
OpKernelConstruction itself contains a status which can be set with TF_OpKernelConstruction_Failure. I'll update the doc.